turbine/src/xdp.rs

```diff
@@ -0,0 +1,167 @@
+// re-export since this is needed at validator startup
+pub use agave_xdp::set_cpu_affinity;
+#[cfg(target_os = "linux")]
+use {
+    agave_xdp::{
+        device::{NetworkDevice, QueueId},
+        load_xdp_program,
+        tx_loop::tx_loop,
+    },
+    crossbeam_channel::TryRecvError,
+    std::{sync::Arc, thread::Builder, time::Duration},
+};
+use {
+    crossbeam_channel::{Sender, TrySendError},
+    solana_ledger::shred,
+    std::{error::Error, net::SocketAddr, thread},
+};
+
+#[derive(Clone, Debug)]
+pub struct XdpConfig {
+    pub interface: Option<String>,
+    pub cpus: Vec<usize>,
+    pub zero_copy: bool,
+    pub channel_cap: usize,
+}
+
+impl XdpConfig {
+    // A nice round number
+    const DEFAULT_CHANNEL_CAP: usize = 1_000_000;
+}
+
+impl Default for XdpConfig {
+    fn default() -> Self {
+        Self {
+            interface: None,
+            cpus: vec![],
+            zero_copy: false,
+            channel_cap: Self::DEFAULT_CHANNEL_CAP,
+        }
+    }
+}
+
+impl XdpConfig {
+    pub fn new(interface: Option<impl Into<String>>, cpus: Vec<usize>, zero_copy: bool) -> Self {
+        Self {
+            interface: interface.map(|s| s.into()),
+            cpus,
+            zero_copy,
+            channel_cap: XdpConfig::DEFAULT_CHANNEL_CAP,
+        }
+    }
+}
+
+pub(crate) struct XdpSender {
+    senders: Vec<Sender<(Vec<SocketAddr>, shred::Payload)>>,
+}
+
+impl XdpSender {
+    #[inline]
+    pub(crate) fn try_send(
+        &self,
+        shred_index: u32,
+        addr: Vec<SocketAddr>,
+        payload: shred::Payload,
+    ) -> Result<(), TrySendError<(Vec<SocketAddr>, shred::Payload)>> {
+        self.senders[shred_index as usize % self.senders.len()].try_send((addr.clone(), payload))
+    }
+}
+
+pub(crate) struct XdpRetransmitter {
+    threads: Vec<thread::JoinHandle<()>>,
+}
+
+impl XdpRetransmitter {
+    #[cfg(not(target_os = "linux"))]
+    pub(crate) fn new(
+        _config: XdpConfig,
+        _src_port: u16,
+    ) -> Result<(Self, XdpSender), Box<dyn Error>> {
```

Comment:
maybe anyhow would be less verbose here (since we are using trait objects anyway).

Reply:
Stop trying to make me use more deps I won't 😂

Reply:
Anyhow is a dep we already use, and it does the same exact thing but better (since Err variant is 1 pointer there not two like for dyn Error).

Reply:
anyhow is for apps, thiserror is for libraries. And nobody cares about two pointers in an error path that won't ever happen. Makes zero sense to add a dependency for this.

Reply:
Well if you intend this code to eventually land back in aya it makes sense. I suggest you add this context to the PR description so other reviewers get a better frame of reference.